eyelid-match-ops/src/primitives/hamming.rs

```diff
@@ -0,0 +1,125 @@
+//! Implementation of the simple encoding
+
+use crate::primitives::poly::Poly;
+use ark_ff::Zero;
+use std::ops::{AddAssign, Sub};
+
+use rand::rngs::ThreadRng;
+
+use super::yashe::Yashe;
+use super::yashe::{Ciphertext, Message, PrivateKey, PublicKey, YasheConf};
+
+/// Contains the message to be encoded such that
+/// the Hamming distance can be computed later.
+pub struct SimpleHammingEncoding<C: YasheConf>
+where
+    C::Coeff: From<u128> + From<u64> + From<i64>,
+{
+    /// The message to be encoded
+    m: Message<C>,
+    /// The reverse of the message to be encoded
+    m_rev: Message<C>,
+}
+
+/// SimpleHammingEncodingCiphertext is a struct that holds two ciphertexts, c and c_rev,
+/// which are the encodings of the message m and m_rev, respectively. The encoding is
+/// done by reversing the message and encoding it as a regular Yashe Ciphertext.
+pub struct SimpleHammingEncodingCiphertext<C: YasheConf>
+where
+    C::Coeff: From<u128> + From<u64> + From<i64>,
+{
+    /// The ciphertext of the message m
+    c: Ciphertext<C>,
+    /// The ciphertext of the message m_rev
+    c_rev: Ciphertext<C>,
+}
+
+impl<C: YasheConf> SimpleHammingEncoding<C>
+where
+    C::Coeff: From<u128> + From<u64> + From<i64>,
+{
+    /// Creates a new `SimpleHammingEncoding` with the given message `m` and size `size`.
+    pub fn new(m: Message<C>, size: usize) -> Self {
+        let mut m_rev = Message {
+            m: Poly::<C>::zero(),
+        };
+        for i in 0..size {
+            m_rev.m[i] = m.m[size - i - 1];
+        }
```

Comment:
I think this manual reverse loop can be implemented as:
```rust
let mut rev = m.m.coeffs().clone();
rev.resize(size, C::Coeff::zero());
rev.reverse();
rev.truncate_to_canonical_form();
```

This is more memory efficient in the common case where all coefficients are non-zero, because it only allocates once. (Rather than `log(MAX_POLY_DEGREE)` times.)

Maybe it's worth having a reverse method on `Poly`?

Reply:
This can become a TODO comment.

Reply:
```suggestion
        // TODO: replace this with coeffs().clone(), Vec resize()/revert(), then truncate_to_canonical_form()
        let mut m_rev = Message {
            m: Poly::<C>::zero(),
        };
        for i in 0..size {
            m_rev.m[i] = m.m[size - i - 1];
        }
```